### PR TITLE
Fix dag dependencies detection

### PR DESCRIPTION
**Pull request by dstandish**
Fix issue where we were creating duplicate deps records in the dag_dependencies node in serialized dag.

Also added some test coverage for the "dependency detector".

cc: @blag

### CODE REVIEW

File: tests/serialization/test_dag_serialization.py [hunk 1]
```diff
@@ -1329,6 +1367,110 @@ class DerivedSensor(ExternalTaskSensor):
...
+        {
+            (
+                'scheduler',
+                'dependency_detector',
+            ): 'tests.serialization.test_dag_serialization.CustomDependencyDetector'
+        }
+    )
+    def test_custom_dep_detector(self):
+        """
+        Prior to deprecation of custom dependency detector, the return type as Optional[DagDependency].
```
Comment (jedcunningham) [inline tests/serialization/test_dag_serialization.py]:
```suggestion
        Prior to deprecation of custom dependency detector, the return type was Optional[DagDependency].
```

File: tests/serialization/test_dag_serialization.py [hunk 2]
```diff
@@ -1329,6 +1367,110 @@ class DerivedSensor(ExternalTaskSensor):
...
+
+        execution_date = datetime(2020, 1, 1)
+        with DAG(dag_id="test", start_date=execution_date) as dag:
+            task1 = ExternalTaskSensor(
+                task_id="task1",
+                external_dag_id="external_dag_id",
+                mode="reschedule",
+            )
+            task2 = EmptyOperator(task_id="task2")
+            task1 >> task2
```
Comment (jedcunningham) [inline tests/serialization/test_dag_serialization.py]:
```suggestion
```

Not sure we need these for this test?

File: tests/serialization/test_dag_serialization.py [hunk 3]
```diff
@@ -1329,6 +1367,110 @@ class DerivedSensor(ExternalTaskSensor):
...
+        d4 = Dataset('d4')
+        execution_date = datetime(2020, 1, 1)
+        with DAG(dag_id="test", start_date=execution_date, schedule_on=[d1]) as dag:
+            task1 = ExternalTaskSensor(
+                task_id="task1",
+                external_dag_id="external_dag_id",
+                mode="reschedule",
+            )
+            task2 = EmptyOperator(task_id="task2")
+            task1 >> task2
```
Comment (jedcunningham) [inline tests/serialization/test_dag_serialization.py]:
```suggestion
```

Same here.

File: tests/serialization/test_dag_serialization.py [hunk 4]
```diff
@@ -1329,6 +1367,110 @@ class DerivedSensor(ExternalTaskSensor):
...
+    def test_dag_deps_datasets(self):
+        """
+        Check that dag_dependencies node is populated correctly for a DAG with datasets.
+        """
+        from airflow.operators.empty import EmptyOperator
+        from airflow.sensors.external_task import ExternalTaskSensor
+
+        d1 = Dataset('d1')
+        d2 = Dataset('d2')
+        d3 = Dataset('d2')
```
Comment (jedcunningham) [inline tests/serialization/test_dag_serialization.py]: Might be worth a different name here, took me a minute to work out the d3.uri _should_ == "d2".

File: airflow/serialization/serialized_objects.py [hunk 5]
```diff
@@ -1048,14 +1058,11 @@ def serialize_dag(cls, dag: DAG) -> dict:
...
-            dag_deps = [
-                t.__dict__
-                for task in dag.task_dict.values()
-                for t in SerializedBaseOperator.detect_dependencies(task)
-                if t is not None
-            ]
-
-            serialized_dag["dag_dependencies"] = dag_deps
+            dag_deps = {
+                d for t in dag.task_dict.values() for d in SerializedBaseOperator.detect_dependencies(t)
```
Comment (jedcunningham) [inline airflow/serialization/serialized_objects.py]:
```suggestion
                dep for task in dag.task_dict.values() for dep in SerializedBaseOperator.detect_dependencies(task)
```

Might be worth using better variable names here.

File: tests/serialization/test_dag_serialization.py (hunk 4 above)
Comment (dstandish) [reply tests/serialization/test_dag_serialization.py]: oh, ... yeah it shouldn't haha that's my b

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**